### PR TITLE
Rectangles includes a measure then layout pass

### PR DESCRIPTION
**Pull request by ColinEberhardt**
@chrisprice - this seems to work quite nicely, by having `rectangles` render the component before invoking `size` and `position` you can measure the added elements directly. I like the way this gives a two pass rendering, measure then layout.

TODO:
- [x] update docs

### CODE REVIEW

File: src/layout/rectangles.js [hunk 1]
```diff
@@ -33,10 +32,15 @@ export default function(layoutStrategy) {
         }
 
         selection.each(function(data, index) {
+
+            var g = dataJoin(this, data);
+            g.call(component);
```
Comment (chrisprice) [inline src/layout/rectangles.js]: Could chain onto previous line?

File: src/layout/rectangles.js [hunk 2]
```diff
@@ -58,7 +62,7 @@ export default function(layoutStrategy) {
             // filter the data to include only those remaining after collision removal
             var filteredData = filteredLayout.map(function(d) { return data[d.dataIndex]; });
 
-            var g = dataJoin(this, filteredData);
+            g = dataJoin(this, filteredData);
```
Comment (chrisprice) [inline src/layout/rectangles.js]: If the filtering only affected `display`/`visibility`/`opacity` of the rectangle, then I don't think there wouldn't be a need for a second join. Do you think that would be a reasonable compromise?

File: src/layout/rectangles.js [hunk 3]
```diff
@@ -58,7 +62,7 @@ export default function(layoutStrategy) {
             // filter the data to include only those remaining after collision removal
             var filteredData = filteredLayout.map(function(d) { return data[d.dataIndex]; });
```
Comment (chrisprice) [inline src/layout/rectangles.js]: I know this is pre-existing but adding the indicies seems a bit weird.

File: src/layout/rectangles.js [hunk 4]
```diff
@@ -33,10 +32,15 @@ export default function(layoutStrategy) {
...
         selection.each(function(data, index) {
+
+            var g = dataJoin(this, data);
+            g.call(component);
+
             // obtain the rectangular bounding boxes for each child
-            var childRects = data.map(function(d, i) {
-                var childPos = position(d, i);
-                var childSize = size(d, i);
+            var childRects = g[0].map(function(node, i) {
```
Comment (chrisprice) [inline src/layout/rectangles.js]: Is this not the equivalent of an `each`?

Comment (chrisprice) [inline src/layout/rectangles.js]: No it's not...

Comment (chrisprice) [inline src/layout/rectangles.js]: It's just the [0] looks odd

File: site/src/components/layout/rectangles-example-measure.js [hunk 5]
```diff
@@ -112,35 +112,20 @@ var data = [
...
-            .attr({
-                // vertically centre the text in the container
-                'dominant-baseline': 'middle',
-                'dy': '-0.5em'
-            });
-    selection.layout();
-}
+var label = fc.tool.container()
+    .padding(labelMargin)
+    .component(function(sel) {
```
Comment (chrisprice) [inline site/src/components/layout/rectangles-example-measure.js]: Did you try decorate instead of component?

Comment (ColinEberhardt) [inline site/src/components/layout/rectangles-example-measure.js]: Yes, I tried decorate - the implementation is pretty similar, although you are piggy-backing the container components data-join, which makes it a bit simpler. However, by using decorate you are becoming tightly coupled to the container components internal structure, you have to know the correct selector to place the `text` in the correct location. This implementation, although slightly more complex, felt more correct.

File: src/layout/rectangles.js (hunk 4 above)
Comment (ColinEberhardt) [inline src/layout/rectangles.js]: That's a good point, I was meaning to re-visit this. I can't think of a better way to map a d3 selection to an array. Have you seen any alternatives?

File: src/layout/rectangles.js (hunk 2 above)
Comment (ColinEberhardt) [inline src/layout/rectangles.js]: That's a good idea, by hiding filtered labels I can avoid a second join - this also means that components that are rendered via this layout can be a little simpler, i.e. they don't need to respect enter / update.

File: site/src/components/layout/rectangles-example-measure.js (hunk 5 above)
Comment (chrisprice) [inline site/src/components/layout/rectangles-example-measure.js]: It seems clearer to me, unless I'm missing something?
```js
.decorate(function(g) {
  g.enter()
    .append('text')
    .attr('dy', '0.7em')
    .text(function(d) { return d.language; });
})
```

Comment (ColinEberhardt) [inline site/src/components/layout/rectangles-example-measure.js]: You're missing one small step ...


```
.decorate(function(g) {
  g.enter()
    .select('g')
    .append('text')
    .attr('dy', '0.7em')
    .text(function(d) { return d.language; });
})
```

You need to select the inner `g` element that has the padding. It's a small thing, but couples this code to the internal structure.

Comment (chrisprice) [inline site/src/components/layout/rectangles-example-measure.js]: Ah fair enough.

File: src/layout/rectangles.js (hunk 4 above)
Comment (chrisprice) [inline src/layout/rectangles.js]: I don't think I have. I suppose in d3 it's unusual to want to do a projection of nodes into data, it's normally the other way around.

**Comment by ColinEberhardt** [comment]
@chrisprice I've updated the component to hide rectangles when filtered, rather than perform the data join twice. Docs updates.

File: site/src/components/layout/rectangles-example-measure.js [hunk 6]
```diff
@@ -158,7 +140,14 @@ var chart = fc.chart.cartesian(
 
 // create the layout that positions the labels
 var labels = fc.layout.rectangles(fc.layout.strategy.greedy())
-        .size(function(d) { return boundingBox(d.language); })
+        .size(function(d) {
+            // measure the label and add the required padding
+            var textSize = d3.select(this)
+                    .select('text')
```
Comment (chrisprice) [inline site/src/components/layout/rectangles-example-measure.js]: Indentation

File: src/layout/rectangles.js [hunk 7]
```diff
@@ -49,33 +53,34 @@ export default function(layoutStrategy) {
             // or number of rectangles.
             var layout = strategy(childRects);
 
-            // add indices to the layouts
+            // add indices to the layouts - this is require in order to determine which
```
Comment (chrisprice) [inline src/layout/rectangles.js]: After discussion -
* [ ] add optional `hidden` flag to objects returned from layout
* [ ] remove `filter` property
* [ ] consolidate `g.attr` calls

**Comment by ColinEberhardt** [comment]
@chrisprice - I've made those changes. I really like the removal of `filter`, which means `removeOverlaps` is a strategy like any other, which means it can now be used on its own:

https://github.com/ScottLogic/d3fc/commit/88dff166cd8e77dce4189f670e6850cfccc40551#diff-dc4b01aaa354fb8a4d962bee3a6d280dR27

**Comment by chrisprice** [comment]
One unrelated comment but otherwise now LGTM.

**Pull request merged**